Add Linux Support (and cmake)

xoorath opened the pull request:
This pull request introduces Linux support with cmake and the windows subsystem for linux (compiling with ninja and gcc).

This PR is to address  #6 and #3

A built version of the mod can be found here: https://github.com/xoorath/FileXT/releases/tag/v1.0.1

Commit:
Removing visual studio project and solution files.

Commit:
Initial cmake setup.

Moves source into a subdirectory.
Introduces a VS2019.bat for generating VS projects.
Documenting required build environment for linux (may change, linux build not hooked up yet)

Commit:
Linux support

Commit:
New VS batch files and update package script

Commit:
Clean script, zipping as part of packaging, warning fixup

Commit:
removing zip from the package process.

Commit:
finished linux support

* log functions are now generic and there are two more variants (LOG_VERBOSE and LOG_CRITICAL) no more numbered log macros
* changed how the storage directory is found and created to fix some order-of-operations crashes
* added a FileXTTest project simply to allow dropping a breakpoint and stepping through (particularly needed for testing linux while developing on windows)

Commit:
Update README.md

xoorath commented:
A built version of the mod can be found here: https://github.com/xoorath/FileXT/releases/tag/v1.0.1

xoorath commented on Arma mod/addons/filext/functions/fn_error.sqf:
```diff
@@ -17,7 +17,8 @@ private _errorText = switch (_errorCode) do {
     case FILEXT_ERROR_WRITE: {"Generic file write error"};
     case FILEXT_ERROR_READ: {"Generic file read error"};
     case FILEXT_ERROR_WRONG_FILE_FORMAT: {"Wrong file format"};
+    case FILEXT_ERROR_BUFFER_TOO_SMALL: {"Buffer too small"};
```
This is for replacing the msvc "_s" variants of functions with an assert.

xoorath commented on FileXT/dllmain.cpp:
```diff
@@ -0,0 +1,347 @@
+bool checkFileName(string& fileName);
+string getDllFolder();
+
+const std::string& GetAndEnsureStorageDir()
```
I tried to avoid refactoring, this change (adding `GetAndEnsureStorageDir`) was to avoid crashes I couldn't effectively debug without a linux development environment. I do like this change, but strictly speaking it shouldn't be necessary for this PR.

xoorath commented on FileXT/value.h:
```diff
@@ -21,6 +21,15 @@
+ * Jan 14 2022 (xoorath): Linux Port
+ *     Changed as<T> functions to named functions (like: asFloat)
+ *     The inline declarations previously used a non-compliant feature of msvc.
+ *     See: https://stackoverflow.com/questions/3052579/explicit-specialization-in-non-namespace-scope/3052604#3052604
```
Note: if we just declare the as functions at namespace scope we will have another issue with the order of function declaration. I think we would have to move practically everything out of class declaration to underneath it or an inl file. I could be wrong, though... this just seemed easier than figuring it out. I don't see an advantage to using a template for as anyways.

Sparker95 replied:
> The inline declarations previously used a non-compliant feature of msvc.

Oh god, what a disaster. Also it's not my library, it was provided by X39. Now looking at its latest version, I think they did same change there:
https://github.com/arma3/sqf-value/blob/master/sqf-value/value.hpp

xoorath replied:
Thanks for linking the file, I actually couldn't find it. Glad to see that author was of the same mind :)

I've updated to use that version of value.hpp instead of the changes I have here.

xoorath commented on FileXTTest/CMakeLists.txt:
```diff
@@ -0,0 +1,25 @@
+    target_link_libraries(FileXTTest FileXT)
+endif (WIN32)
+if (UNIX)
+    target_link_libraries(FileXTTest FileXT dl)
```
dl is needed specifically for the dladdr function. Funny enough we only need this on the executable project, not FileXT.

Commit:
update value.h, turn off verbose, add build_all

Removed my modification of value.h to update to the head version version:  https://github.com/arma3/sqf-value/blob/master/sqf-value/value.hpp

Turnd off verbose logs except in debug builds

Fixed the gitignore to no longer exclude Build_All.bat

Commit:
documenting -debug switch

Sparker95 commented:
The build works on my PC perfectly, so I finally merge it. Thanks a lot!
